File: androidApp/features/notedetail/src/main/kotlin/com/velkonost/getbetter/android/features/notedetail/NoteDetailScreen.kt
package com.velkonost.getbetter.android.features.notedetail

import androidx.compose.animation.AnimatedVisibility
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.PaddingValues
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.material.ExperimentalMaterialApi
import androidx.compose.material.ModalBottomSheetValue
import androidx.compose.material.rememberModalBottomSheetState
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.rememberCoroutineScope
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.unit.dp
import androidx.lifecycle.compose.collectAsStateWithLifecycle
import com.velkonost.getbetter.android.features.areadetail.AreaDetailScreen
import com.velkonost.getbetter.android.features.notedetail.components.ActionButtons
import com.velkonost.getbetter.android.features.notedetail.components.AuthorData
import com.velkonost.getbetter.android.features.notedetail.components.NoteDetailHeader
import com.velkonost.getbetter.android.features.notedetail.components.comments.CommentItem
import com.velkonost.getbetter.android.features.notedetail.components.comments.NewCommentTextField
import com.velkonost.getbetter.android.features.profiledetail.ProfileDetailScreen
import com.velkonost.getbetter.core.compose.components.AppAlertDialog
import com.velkonost.getbetter.core.compose.components.Loader
import com.velkonost.getbetter.core.compose.components.MultilineTextField
import com.velkonost.getbetter.core.compose.components.details.AreaData
import com.velkonost.getbetter.core.compose.components.details.TaskData
import com.velkonost.getbetter.core.compose.components.note.completiondate.CompletionDateBlock
import com.velkonost.getbetter.core.compose.components.note.subnotes.SubNotesBlock
import com.velkonost.getbetter.core.compose.components.note.tags.TagsBlock
import com.velkonost.getbetter.shared.core.model.note.NoteType
import com.velkonost.getbetter.shared.features.notedetail.presentation.NoteDetailViewModel
import com.velkonost.getbetter.shared.features.notedetail.presentation.contract.NavigateBack
import com.velkonost.getbetter.shared.features.notedetail.presentation.contract.NoteDetailAction
import com.velkonost.getbetter.shared.features.notedetail.presentation.contract.NoteDetailEvent
import com.velkonost.getbetter.shared.features.notedetail.presentation.contract.NoteState
import com.velkonost.getbetter.shared.resources.SharedR
import dev.icerock.moko.resources.compose.colorResource
import dev.icerock.moko.resources.compose.stringResource
import kotlinx.coroutines.flow.collectLatest
import kotlinx.coroutines.launch

@OptIn(ExperimentalMaterialApi::class)
@Composable
fun NoteDetailScreen(
    modifier: Modifier = Modifier,
    viewModel: NoteDetailViewModel
) {

    val state by viewModel.viewState.collectAsStateWithLifecycle()
    val context = LocalContext.current
    val scope = rememberCoroutineScope()
    val isSubNotesBlockVisible = remember { mutableStateOf(true) }

    val confirmDeleteNoteDialog = remember { mutableStateOf(false) }
    val areaDetailSheetState = rememberModalBottomSheetState(
        initialValue = ModalBottomSheetValue.Hidden,
        skipHalfExpanded = true
    )
    val profileDetailSheetState = rememberModalBottomSheetState(
        initialValue = ModalBottomSheetValue.Hidden,
        skipHalfExpanded = true
    )

    val selectedAreaId = remember { mutableStateOf<Int?>(null) }
    val selectedUserId = remember { mutableStateOf<String?>(null) }

    Box(modifier = modifier.fillMaxSize()) {
        if (state.isLoading) {
            Loader(modifier = Modifier.align(Alignment.Center))
        } else {
            LazyColumn(
                modifier = modifier
                    .padding(start = 20.dp, end = 20.dp)
                    .fillMaxSize(),
                contentPadding = PaddingValues(bottom = 220.dp)
            ) {
                item {
                    NoteDetailHeader(
                        noteType = state.noteType,
                        isNotePrivate = state.isNotePrivate,
                        likesData = state.likesData,
                        onBackClick = {
                            viewModel.dispatch(NavigateBack)
                        },
                        onLikeClick = {
                            viewModel.dispatch(NoteDetailAction.LikeClick)
                        }
                    )
                }

                if (!state.allowEdit) {
                    item {
                        AuthorData(
                            isLoading = state.authorLoading,
                            author = state.author,
                            onClick = {
                                scope.launch {
                                    selectedUserId.value = state.author?.id
                                    profileDetailSheetState.show()
                                }
                            }
                        )
                    }
                }

                item {
                    state.area?.let {
                        AreaData(area = it) {
                            scope.launch {
                                selectedAreaId.value = state.area!!.id
                                areaDetailSheetState.show()
                            }
                        }
                    }
                }

                state.task?.let {
                    item {
                        TaskData(task = it) {
                            viewModel.dispatch(NoteDetailAction.TaskClick)
                        }
                    }
                }

                item {
                    MultilineTextField(
                        value = state.noteText,
                        placeholderText = stringResource(
                            resource =
                            if (state.noteType == NoteType.Default) SharedR.strings.create_note_text_hint
                            else SharedR.strings.create_goal_text_hint
                        ),
                        isEnabled = state.noteState == NoteState.Editing,
                        onValueChanged = {
                            viewModel.dispatch(NoteDetailAction.TextChanged(it))
                        }
                    )
                }

                item {
                    AnimatedVisibility(
                        visible = state.noteType == NoteType.Goal
                                && (state.allowEdit || state.expectedCompletionDate != null || state.completionDate != null)
                    ) {
                        CompletionDateBlock(
                            enabled = state.noteState == NoteState.Editing,
                            isLoading = state.isCompleteGoalLoading,
                            initialValue = state.expectedCompletionDate,
                            initialValueStr = state.expectedCompletionDateStr?.toString(context),
                            isCompleteVisible = state.allowEdit,
                            completionDateStr = state.completionDateStr?.toString(context),
                            onSetCompletionDate = {
                                viewModel.dispatch(NoteDetailAction.SetCompletionDate(it))
                            },
                            onCompleteClick = {
                                if (state.completionDate == null) {
                                    viewModel.dispatch(NoteDetailAction.CompleteClick)
                                } else {
                                    viewModel.dispatch(NoteDetailAction.UnCompleteClick)
                                }
                            }
                        )
                    }
                }

                item {
                    AnimatedVisibility(
                        visible = state.tags.isNotEmpty() || state.noteState == NoteState.Editing
                    ) {
                        TagsBlock(
                            tags = state.tags,
                            newTag = state.newTag,
                            onlyView = state.noteState == NoteState.View,
                            onNewTagChanged = {
                                viewModel.dispatch(NoteDetailAction.NewTagTextChanged(it))
                            },
                            onAddNewTag = {
                                viewModel.dispatch(NoteDetailAction.AddNewTag)
                            },
                            onTagDelete = {
                                viewModel.dispatch(NoteDetailAction.RemoveTag(it))
                            }
                        )
                    }
                }

                item {
                    AnimatedVisibility(
                        visible = state.noteType == NoteType.Goal
                                && (state.subNotes.isNotEmpty() || state.noteState == NoteState.Editing)
                    ) {
                        SubNotesBlock(
                            items = state.subNotes,
                            newSubNote = state.newSubNote,
                            onlyView = state.noteState == NoteState.View,
                            isCompleteVisible = state.allowEdit,
                            isSubNotesBlockVisible = isSubNotesBlockVisible,
                            onAddNewSubNote = {
                                viewModel.dispatch(NoteDetailAction.AddSubNote)
                            },
                            onNewSubNoteChanged = {
                                viewModel.dispatch(NoteDetailAction.NewSubNoteTextChanged(it))
                            },
                            onSubNoteDelete = {
                                viewModel.dispatch(NoteDetailAction.RemoveSubNote(it))
                            },
                            onCompleteClick = {
                                if (it.completionDate == null) {
                                    viewModel.dispatch(NoteDetailAction.CompleteSubNoteClick(it))
                                } else {
                                    viewModel.dispatch(NoteDetailAction.UnCompleteSubNoteClick(it))
                                }
                            }
                        )
                    }
                }

                item {
                    AnimatedVisibility(
                        modifier = modifier.padding(top = 12.dp),
                        visible = state.allowEdit
                    ) {
                        ActionButtons(
                            noteState = state.noteState,
                            onEditClick = {
                                viewModel.dispatch(NoteDetailAction.StartEditClick)
                            },
                            onDeleteClick = {
                                confirmDeleteNoteDialog.value = true

                            },
                            onSaveClick = {
                                viewModel.dispatch(NoteDetailAction.EndEditClick)
                            },
                            onCancelSaveClick = {
                                viewModel.dispatch(NoteDetailAction.CancelEditClick)
                            }
                        )
                    }
                }

                item {
                    AnimatedVisibility(visible = state.commentsData.comments.isNotEmpty()) {
                        Text(
                            modifier = modifier
                                .padding(top = 24.dp)
                                .fillMaxWidth(0.8f),
                            text = stringResource(resource = SharedR.strings.note_detail_comments_title),
                            color = colorResource(resource = SharedR.colors.text_primary),
                            style = MaterialTheme.typography.headlineSmall
                        )
                    }
                }

                state.commentsData.comments.forEach {
                    item {
                        CommentItem(
                            item = it,
                            onDeleteClick = {
                                viewModel.dispatch(NoteDetailAction.CommentRemoveClick(it))
                            }
                        )

                    }
                }
            }
        }
    }

    if (!state.isNotePrivate) {
        Column(modifier = modifier.fillMaxSize()) {
            Spacer(modifier.weight(1f))
            NewCommentTextField(
                value = state.commentsData.commentText,
                onValueChanged = {
                    viewModel.dispatch(NoteDetailAction.CommentTextChanged(it))
                },
                onSendClick = {
                    viewModel.dispatch(NoteDetailAction.CommentAddClick)
                }
            )
        }
    }

    AreaDetailScreen(
        modalSheetState = areaDetailSheetState,
        areaId = selectedAreaId.value,
        onAreaChanged = {
            viewModel.dispatch(NoteDetailAction.AreaChanged)
        }
    )

    ProfileDetailScreen(
        modalSheetState = profileDetailSheetState,
        userId = selectedUserId.value
    )

    if (confirmDeleteNoteDialog.value) {
        AppAlertDialog(
            title = stringResource(
                resource =
                if (state.noteType == NoteType.Default) SharedR.strings.note_detail_confirm_delete_title
                else SharedR.strings.goal_detail_confirm_delete_title
            ),
            text = stringResource(resource = SharedR.strings.note_detail_confirm_delete_text),
            confirmTitle = stringResource(resource = SharedR.strings.confirm),
            cancelTitle = stringResource(resource = SharedR.strings.cancel),
            onDismiss = { confirmDeleteNoteDialog.value = false },
            onConfirmClick = {
                viewModel.dispatch(NoteDetailAction.DeleteClick)
                confirmDeleteNoteDialog.value = false
            }
        )
    }

    LaunchedEffect(Unit) {
        viewModel.events.collectLatest {
            when (it) {
                is NoteDetailEvent.DeleteSuccess -> {
                    viewModel.dispatch(NavigateBack)
                }

                is NoteDetailEvent.EditSuccess -> {
                    //TODO
                }
            }
        }
    }

}

File: iosApp/iosApp/Features/NoteDetail/NoteDetailScreen.swift
//
//  NoteDetailScreen.swift
//  iosApp
//
//  Created by velkonost on 03.11.2023.
//  Copyright © 2023 orgName. All rights reserved.
//

import Foundation
import SwiftUI
import SharedSDK
import KMMViewModelSwiftUI
import KMPNativeCoroutinesAsync

struct NoteDetailScreen : View {
    
    @Environment(\.presentationMode) var presentationMode: Binding<PresentationMode>
    
    @StateViewModel var viewModel: NoteDetailViewModel
    @State private var eventsObserver: Task<(), Error>? = nil
    
    @State private var confirmDeleteNoteDialog = false
    @State private var isSubNotesBlockVisible = true
    
    @State private var selectedAreaId: Int32? = nil
    @State private var selectedUserId: String? = nil
    
    @State private var showingAreaDetailSheet = false
    @State private var showingProfileDetailSheet = false
    
    var body: some View {
        @State var state = viewModel.viewStateValue as! NoteDetailViewState
        @State var newTag = state.newTag
        @State var newSubNote = state.newSubNote
        @State var noteText = state.noteText
        
        ZStack {
            Color.mainBackground
            
            if state.isLoading {
                VStack {
                    Loader()
                }
            } else {
                ScrollView(.vertical, showsIndicators: false) {
                    ScrollViewReader { value in
                        
                        LazyVStack {
                            NoteDetailHeader(
                                noteType: state.noteType,
                                isNotePrivate: state.isNotePrivate,
                                likesData: state.likesData,
                                onLikeClick: {
                                    viewModel.dispatch(action: NoteDetailActionLikeClick())
                                }
                            )
                            
                            if !state.allowEdit {
                                AuthorData(
                                    isLoading: state.authorLoading,
                                    author: state.author,
                                    onClick: {
                                        if state.author?.id != nil {
                                            selectedUserId = state.author?.id
                                            showingProfileDetailSheet = true
                                        }
                                    }
                                )
                            }
                            
                            if state.area != nil {
                                AreaData(
                                    area: state.area!,
                                    onClick: {
                                        selectedAreaId = state.area!.id
                                        showingAreaDetailSheet = true
                                    }
                                )
                            }
                            
                            if state.task != nil {
                                TaskData(
                                    task: state.task!,
                                    onClick: {
                                        viewModel.dispatch(action: NoteDetailActionTaskClick())
                                    }
                                )
                            }
                            
                            if !noteText.isEmpty {
                                MultilineTextFieldBinding(
                                    value: $noteText,
                                    placeholderText:
                                        state.noteType == NoteType.default_ ? SharedR.strings().create_note_text_hint.desc().localized() : SharedR.strings().create_goal_text_hint.desc().localized(),
                                    isEnabled: state.noteState == NoteState.editing
                                ) { value in
                                    viewModel.dispatch(action: NoteDetailActionTextChanged(value: value))
                                }
                                .onTapGesture {
                                    withAnimation {
                                        value.scrollTo(1, anchor: .bottom)
                                    }
                                }
                                .id(1)
                            }
                            
                            if state.noteType == NoteType.goal && (state.allowEdit || state.expectedCompletionDate != nil || state.completionDate != nil) {
                                CompletionDateBlock(
                                    enabled: state.noteState == NoteState.editing,
                                    isLoading: state.isCompleteGoalLoading,
                                    initialValue: state.expectedCompletionDate as? Int64,
                                    initialValueStr: state.expectedCompletionDateStr?.localized(),
                                    isCompleteVisible: state.allowEdit,
                                    completionDateStr: state.completionDateStr?.localized(),
                                    onSetCompletionDate: { value in
                                        viewModel.dispatch(action: NoteDetailActionSetCompletionDate(value: KotlinLong(value: value!)))
                                    },
                                    onCompleteClick: {
                                        if state.completionDate == nil {
                                            viewModel.dispatch(action: NoteDetailActionCompleteClick())
                                        } else {
                                            viewModel.dispatch(action: NoteDetailActionUnCompleteClick())
                                        }
                                    }
                                )
                            }
                            
                            if !state.tags.isEmpty || state.noteState == NoteState.editing {
                                TagsBlock(
                                    tags: state.tags,
                                    newTag: $newTag,
                                    onlyView: state.noteState == NoteState.view,
                                    onNewTagChanged: { value in
                                        viewModel.dispatch(action: NoteDetailActionNewTagTextChanged(value: value))
                                    },
                                    onAddNewTag: {
                                        viewModel.dispatch(action: NoteDetailActionAddNewTag())
                                    },
                                    onTagDelete: { value in
                                        viewModel.dispatch(action: NoteDetailActionRemoveTag(value: value))
                                    }
                                )
                                .onTapGesture {
                                    withAnimation {
                                        value.scrollTo(2, anchor: .bottom)
                                    }
                                }
                                .id(2)
                            }
                            
                            if state.noteType == NoteType.goal && (!state.subNotes.isEmpty || state.noteState == NoteState.editing) {
                                SubNotesBlock(
                                    items: state.subNotes,
                                    newSubNote: $newSubNote,
                                    onlyView: state.noteState == NoteState.view,
                                    isCompleteVisible: state.allowEdit,
                                    onNewSubNoteChanged: { value in
                                        viewModel.dispatch(action: NoteDetailActionNewSubNoteTextChanged(value: value))
                                    },
                                    onAddNewSubNote: {
                                        viewModel.dispatch(action: NoteDetailActionAddSubNote())
                                    },
                                    onSubNoteDelete: { value in
                                        viewModel.dispatch(action: NoteDetailActionRemoveSubNote(value: value))
                                    },
                                    isSubNotesBlockPickerVisible: $isSubNotesBlockVisible,
                                    onCompleteClick: { value in
                                        if value.completionDate == nil {
                                            viewModel.dispatch(action: NoteDetailActionCompleteSubNoteClick(value: value))
                                        } else {
                                            viewModel.dispatch(action: NoteDetailActionUnCompleteSubNoteClick(value: value))
                                        }
                                    }
                                )
                                .onTapGesture {
                                    withAnimation {
                                        value.scrollTo(3, anchor: .bottom)
                                    }
                                }
                                .id(3)
                            }
                            
                            if state.allowEdit {
                                ActionButtons(
                                    noteState: state.noteState,
                                    onEditClick: {
                                        viewModel.dispatch(action: NoteDetailActionStartEditClick())
                                    },
                                    onDeleteClick: {
                                        confirmDeleteNoteDialog = true
                                    },
                                    onSaveClick: {
                                        viewModel.dispatch(action: NoteDetailActionEndEditClick())
                                    },
                                    onCancelSaveClick: {
                                        viewModel.dispatch(action: NoteDetailActionCancelEditClick())
                                    }
                                )
                            }
                            
                            if !state.commentsData.comments.isEmpty {
                                HStack {
                                    Text(SharedR.strings().note_detail_comments_title.desc().localized())
                                        .style(.headlineSmall)
                                        .foregroundColor(.textPrimary)
                                    Spacer()
                                }
                            }
                            
                            ForEach(state.commentsData.comments, id: \.self.id) { item in
                                CommentItem(
                                    item: item,
                                    onDeleteClick: { value in
                                        viewModel.dispatch(action: NoteDetailActionCommentRemoveClick(value: value))
                                    }
                                )
                            }.id(UUID())
                            
                            Spacer().frame(height: 140)
                            
                        }
                        .padding(.horizontal, 20)
                    }
                }
                .onTapGesture {
                    endTextEditing()
                }
                
                if !state.isNotePrivate {
                    VStack(spacing: 0) {
                        Spacer()
                        NewCommentTextField(
                            value: Binding(get: { state.commentsData.commentText }, set: { value in }),
                            onValueChanged: { value in
                                viewModel.dispatch(action: NoteDetailActionCommentTextChanged(value: value))
                            },
                            onSendClick: {
                                viewModel.dispatch(action: NoteDetailActionCommentAddClick())
                            }
                        )
                    }
                    .ignoresSafeArea(.container, edges: .bottom)
                }
            }
            
        }
        .alert(
            state.noteType == NoteType.default_ ? SharedR.strings().note_detail_confirm_delete_title.desc().localized() : SharedR.strings().goal_detail_confirm_delete_title.desc().localized(), isPresented: $confirmDeleteNoteDialog) {
                Button(SharedR.strings().confirm.desc().localized()) {
                    viewModel.dispatch(action: NoteDetailActionDeleteClick())
                }
                Button(SharedR.strings().cancel.desc().localized(), role: .cancel) {}
            } message: {
                Text(SharedR.strings().note_detail_confirm_delete_text.desc().localized())
            }
        
            .sheet(isPresented: $showingAreaDetailSheet) {
                AreaDetailScreen(
                    areaId: $selectedAreaId,
                    onClose: {
                        showingAreaDetailSheet = false
                    },
                    onAreaChanged: { areaId in
                        viewModel.dispatch(action: NoteDetailActionAreaChanged())
                    }
                )
            }
            .sheet(isPresented: $showingProfileDetailSheet) {
                ProfileDetailScreen(userId: $selectedUserId)
            }
            .onAppear {
                observeEvents()
            }
    }
}

extension NoteDetailScreen {
    func observeEvents() {
        if eventsObserver == nil {
            eventsObserver = Task {
                for try await event in asyncSequence(for: viewModel.events) {
                    switch(event) {
                    case _ as NoteDetailEventDeleteSuccess: do {
                        presentationMode.wrappedValue.dismiss()
                    }
                        
                    case _ as NoteDetailEventEditSuccess: do {
                        // TODO
                    }
                        
                        
                    default:
                        break
                    }
                }
            }
        }
    }
    
}
